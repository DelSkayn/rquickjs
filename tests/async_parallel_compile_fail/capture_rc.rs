use std::rc::Rc;

use rquickjs::{async_with, AsyncContext, AsyncRuntime};

pub async fn test() {
    let rt = AsyncRuntime::new().unwrap();
    let ctx = AsyncContext::full(&rt).await.unwrap();

    let fut = {
        let mut var = Rc::new(1);
        let var_c = var.clone();
        async_with!(ctx => |_ctx|{
            // you should not be able to move non send types into the closure.
            assert_eq!(*var_c,1);
        })
    };
    fut.await;
}

fn main() {}
