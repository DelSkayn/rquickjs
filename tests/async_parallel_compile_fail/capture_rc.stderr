error[E0277]: `Rc<i32>` cannot be sent between threads safely
  --> tests/async_parallel_compile_fail/capture_rc.rs:12:9
   |
12 | /         async_with!(ctx => |_ctx|{
13 | |             // you should not be able to move non send types into the closure.
14 | |             assert_eq!(*var_c,1);
15 | |         })
   | |          ^
   | |          |
   | |          `Rc<i32>` cannot be sent between threads safely
   | |__________within this `[closure@$DIR/src/context/async.rs:106:52: 106:58]`
   |            required by a bound introduced by this call
   |
   = help: within `[closure@$DIR/src/context/async.rs:106:52: 106:58]`, the trait `Send` is not implemented for `Rc<i32>`
note: required because it's used within this closure
  --> tests/async_parallel_compile_fail/capture_rc.rs:12:9
   |
12 | /         async_with!(ctx => |_ctx|{
13 | |             // you should not be able to move non send types into the closure.
14 | |             assert_eq!(*var_c,1);
15 | |         })
   | |__________^
   = note: required for `[closure@$DIR/src/context/async.rs:106:52: 106:58]` to implement `ParallelSend`
note: required by a bound in `AsyncContext::async_with`
  --> src/context/async.rs
   |
   |     pub async fn async_with<'a, F, R: 'a>(&'a self, f: F) -> R
   |                  ---------- required by a bound in this associated function
...
   |             + ParallelSend,
   |               ^^^^^^^^^^^^ required by this bound in `AsyncContext::async_with`
   = note: this error originates in the macro `async_with` (in Nightly builds, run with -Z macro-backtrace for more info)
